### ISSUE TITLE
Proper multi-parameter functions

### ISSUE THREAD

**Issue by Athas** [HIPERFIT/futhark#228]
Our function declaration syntax is now nice, but underneath it's as simplistic as it has always been.  Writing Futhark functions that accept tuples as arguments is awkward, because you cannot destructure them in the pattern itself.

We need proper multi-parameter functions, and we need proper destructuring patterns.  In Futhark-as-of-now, this function has two parameters:

```
fun f (x: int, y: int): int = x + y
```

In Futhark-of-the-future, `f` will be a unary function that accepts a tuple.  It could then also be written as:

```
fun f ((x,y) : (int,int)): int = x + y
```

What is interesting is that we can then define this differently-typed function, which would be a type error in Futhark-as-of-now:

```
fun g (x: int) (y: int): int = x + y
```

This change will allow us to have a more sensible `zipWith`, as well as get rid of the horrible "tuple shimming" hack that we use to make `map` and `zip` work with minimal notational overhead.  This change will affect only the source language - the core language doesn't have tuples, after all.  There is nothing novel or difficult here; just integrating a 70s language feature and being Futhark ever closer to being an ML.

**Comment by Athas** [comment]
Also, we won't implement actual pattern matching for now, just simple tuple destructuring.   Branches are performance poison, and we gotta go fast.

**Comment by Athas** [comment]
I realised this will also require Futhark to change its function application syntax to be juxtaposition, just like every other functional language.

**Comment by Athas** [comment]
Closed in 50f2c66b1a549b06e7c2b3d8cc265c3cdf539af4.  This turned out to be a major language change.  We will have to make more changes in the future to polish this (for example to the SOAC syntax).
